ecs-init/volumes/ecs_volume_plugin.go

```diff
@@ -201,32 +188,148 @@ func deleteMountPath(path string) error {
 
 // Mount implements Docker volume plugin's Mount Method
 func (a *AmazonECSVolumePlugin) Mount(r *volume.MountRequest) (*volume.MountResponse, error) {
-	a.lock.RLock()
-	defer a.lock.RUnlock()
+	seelog.Infof("Received mount request %+v", r)
+
+	// Validate the request
+	if len(r.Name) == 0 {
+		return nil, fmt.Errorf("no volume in the request")
+	}
+	if len(r.ID) == 0 {
+		return nil, fmt.Errorf("no mount ID in the request")
+	}
+
+	// Acquire write lock
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	// Find the volume
 	vol, ok := a.volumes[r.Name]
 	if !ok {
 		seelog.Errorf("Volume %s to mount is not found", r.Name)
 		return nil, fmt.Errorf("volume %s not found", r.Name)
 	}
+
+	// Find the volume driver
+	volDriver, err := a.getVolumeDriver(vol.Type)
+	if err != nil {
+		seelog.Errorf("Volume %s's driver type %s not supported: %v", r.Name, vol.Type, err)
+		return nil, fmt.Errorf("Volume %s's driver type %s not supported: %w", r.Name, vol.Type, err)
+	}
+	if volDriver == nil {
+		// This case shouldn't happen normally
+		return nil, fmt.Errorf("no volume driver found for type %s", vol.Type)
+	}
+
+	// Mount the volume on the host if there are no active mounts for the volume.
+	if len(vol.Mounts) == 0 {
+		seelog.Infof("Mounting volume %s as there are no existing mounts for it", r.Name)
+		createReq := &driver.CreateRequest{Name: r.Name, Path: vol.Path, Options: vol.Options}
+		if err := volDriver.Create(createReq); err != nil {
+			seelog.Errorf("Volume %s creation failure: %v", r.Name, err)
+			return nil, fmt.Errorf("failed to mount volume %s: %w", r.Name, err)
+		}
+		seelog.Infof("Volume %s mounted successfully", r.Name)
+	}
+
+	// Update state
+	seelog.Infof("Adding mount %s to volume %s", r.ID, r.Name)
+	vol.AddMount(r.ID)
+	if err := a.state.recordVolume(r.Name, vol); err != nil {
+		// State update failed, so roll back the changes made so far to make state consistent
+		seelog.Errorf("Failed to save volume %s, rolling back changes: %v", r.Name, err)
+		vol.RemoveMount(r.ID)
+		if len(vol.Mounts) == 0 {
+			seelog.Warnf("Rolling back mounting of volume %s", r.Name)
+			if err := volDriver.Remove(&driver.RemoveRequest{Name: r.Name}); err != nil {
+				seelog.Errorf("Volume %s removal failure: %v", r.Name, err)
+			}
+		}
+		a.state.recordVolume(r.Name, vol)
```

Comment:
Maybe we don't need this line if the known status has failed to be updated in the first place?

Reply:
`a.state` is stateful and maintains a copy of the volumes in a `map`. When it fails to save the volumes to the file it simply returns the error to the caller without correcting its in-memory state. This call is to make `a.state`'s in-memory state consistent with `a`'s state. 

TBH `a.state` should be stateless and doesn't have an in-memory state at all. Same is true for `ECSVolumeDriver` struct which also has an unnecessary in-memory state where it keeps the volumes. I will add a task to make these two components stateless.